### PR TITLE
Add "local" log driver

### PR DESCRIPTION
**Pull request by cpuguy83**
This adds the "local" log driver described in #33475.

### CODE REVIEW

File: daemon/logger/jsonfilelog/read.go
Comment (anusha-ragunathan) [inline daemon/logger/jsonfilelog/read.go]: Why is retries not incremented? Both here and L73?

File: daemon/logger/loggerutils/logfile_test.go
Comment (anusha-ragunathan) [inline daemon/logger/loggerutils/logfile_test.go]: We should add s4 with `Why does it say paper jam when there is no paper jam?`.

**Comment by anusha-ragunathan** [comment]
Overall LGTM.

Needs a followup doc PR

File: daemon/logger/jsonfilelog/read.go
Comment (cpuguy83) [reply daemon/logger/jsonfilelog/read.go]: These were erroneous. The for loop already increments retries.

File: daemon/logger/jsonfilelog/jsonfilelog_test.go
Comment (thaJeztah) [inline daemon/logger/jsonfilelog/jsonfilelog_test.go]: IIUC, the intent of this logging driver is to be an "optimised, general-purpose logging driver" (which could become the default).

The storage mechanisms for this driver (currently, file-based, file-format-x) is an implementation detail. Because of that, should we abstract these options? i.e., have options for:

- "maximum amount of logs to keep" (the driver can decide to use multiple files to optimize)
- "maximum age of logs to keep" (in light of GDPR https://github.com/moby/moby/issues/36853)
- compression; good one: is there a significant overhead? are there situations where you _want_ to disable this?

File: daemon/logger/local/doc.go
Comment (thaJeztah) [inline daemon/logger/local/doc.go]: Q: If I understand correctly, we want the storage-format to be an implementation detail; if we ever decide to switch to/add a new storage format, would there be a need to store the format in the container's configuration?

**Review by thaJeztah — CHANGES_REQUESTED** [review]

File: daemon/logger/local/doc.go
Comment (cpuguy83) [reply daemon/logger/local/doc.go]: This is covered a bit in the propsoal issue.
Basically we'd store the format details in the header of the log file.

I started to go ahead and add support in `logfile` for setting headers, but it's a significant change, and technically not needed for the first format since we can detect that no format is specified and therefore it's the first format.

Comment (thaJeztah) [reply daemon/logger/local/doc.go]: > technically not needed for the first format since we can detect that no format is specified and therefore it's the first format.

Makes sense, and yes, that would work

**Comment by cpuguy83** [comment]
Moved this to code review per discussion with @thaJeztah on #maintainers in Slack.

File: container/container.go
Comment (vdemeester) [inline container/container.go]: why don't we set `container.LogPath` here too ? :thinking:

**Review by vdemeester — APPROVED** [review]
SGTM :camel:

File: container/container.go
Comment (cpuguy83) [reply container/container.go]: We don't want to expose this to users.

File: daemon/logger/local/config.go
Comment (thaJeztah) [inline daemon/logger/local/config.go]: Per the above; `MaxFileCount == 1` looks to do exactly the same as `MaxFileCount == 0`, so this should probably be;

```go
if cfg.MaxFileCount <= 1 {
```

File: daemon/logger/local/config.go
Comment (thaJeztah) [reply daemon/logger/local/config.go]: Some further testing; if _no_ `max-file` is provided, the default is 4 (5?);

```
Every 2.0s: ls -lah /var/lib/docker/containers/dfc64e17d8fa30ea9d5328d35a28cbe38b3852d69719f12af182a1418c5ce746/local-logs/                                                            117a4e8e575a: Wed Jul 11 12:56:21 2018

total 112K
drwx------ 2 root root 4.0K Jul 11 12:56 .
drwx------ 5 root root 4.0K Jul 11 12:55 ..
-rw-r----- 1 root root  54K Jul 11 12:56 container.log
-rw-r----- 1 root root 9.0K Jul 11 12:56 container.log.1.gz
-rw-r----- 1 root root 9.0K Jul 11 12:56 container.log.2.gz
-rw-r----- 1 root root 8.8K Jul 11 12:56 container.log.3.gz
-rw-r----- 1 root root 9.1K Jul 11 12:56 container.log.4.gz
```

But inspecting the container won't reflect that; should we show the default?

```bash
docker inspect $CID --format '{{json .HostConfig.LogConfig}}'
{"Type":"local","Config":{"max-size":"100k"}}
```

**Review by thaJeztah — CHANGES_REQUESTED** [review]
Finally able to give this another look; found some issues in tests, and some possible issues around `max-file` validation/checks

**Comment by kolyshkin** [comment]
SGTM

File: daemon/logger/local/config.go
Comment (cpuguy83) [reply daemon/logger/local/config.go]: We could... maybe adding a `DefaultConfig()` function to the logger interface?

**Comment by thaJeztah** [comment]
@cpuguy83 build is failing

```
04:15:27 daemon/logger/local/local_test.go:23:2:warning: could not import github.com/gotestyourself/gotestyourself/assert (cannot find package "github.com/gotestyourself/gotestyourself/assert" in any of: (gosimple)
```

File: daemon/logger/local/local_test.go
Comment (thaJeztah) [inline daemon/logger/local/local_test.go]: This should be 

```go
"gotest.tools/assert
```

File: daemon/logger/local/local_test.go
Comment (thaJeztah) [inline daemon/logger/local/local_test.go]:
```go
"gotest.tools/assert/cmp"
```

File: daemon/logger/local/config.go
Comment (cpuguy83) [reply daemon/logger/local/config.go]: Actually I think it would be better to not do this for now at least.
I'd prefer to not change the hostconfig on the container.

File: daemon/logger/local/local_test.go
Comment (cpuguy83) [reply daemon/logger/local/local_test.go]: fixed

File: daemon/logger/local/local_test.go
Comment (cpuguy83) [reply daemon/logger/local/local_test.go]: fixed

**Comment by cpuguy83** [comment]
swagger-gen 👊

**Review by thaJeztah — APPROVED** [review]
LGTM if CI goes green

**Comment by andrewhsu** [comment]
Previous z build [failure](https://jenkins.dockerproject.org/job/Docker-PRs-s390x/10768/console) seems infrastructure network related:
```
00:00:51.155 docker: Error response from daemon: Get https://registry-1.docker.io/v2/library/docker-s390x/manifests/0e10011: net/http: TLS handshake timeout.
```

I kicked the z builds again, but if it can't get a decent run, I propose we ignore z results. We've been having intermittent z infrastructure issues the last few days.

**Comment by thaJeztah** [comment]
Agreed; z failure is definitely not related; everything else is green, so let's get this merged

**Comment by yosifkit** [comment]
Just to be clear for users that read the description in #33475 and think that more is done than is currently complete. It is correct that this PR only addresses the adding of the `local` log driver and does not yet add the "dual-logging" described there?

Maybe add a checklist over there to show when things are merged and which docker release is targeted for changing the default from jsonfile?

**Comment by cpuguy83** [comment]
@yosifkit That is correct, this is only the logger.

**Comment by mterron** [comment]
How do we enable this driver?

**Comment by cpuguy83** [comment]
@mterron it's not in a release yet, but it would be "--log-driver=local"

**Comment by thaJeztah** [comment]
This driver shipped with Docker CE 18.09 and up, for those that were still waiting for this to arrive 👍 (I think it was already included in Docker Enterprise (EE) 18.03)

**Pull request merged**
